contracts/oracle/PCVOracle.sol

```diff
@@ -0,0 +1,370 @@
+            uint256 liquidPcv,
+            uint256 illiquidPcv,
+            uint256 totalPcv
+        )
+    {
+        uint256 liquidVenueLength = liquidVenues.length();
+        uint256 illiquidVenueLength = illiquidVenues.length();
+
+        /// there will never be more than 100 total venues
+        /// so keep the math unchecked to save on gas
+        unchecked {
+            for (uint256 i = 0; i < liquidVenueLength; i++) {
+                address depositAddress = liquidVenues.at(i);
+                (Decimal.D256 memory oracleValue, bool oracleValid) = IOracle(
+                    tokenToOracle[depositToToken[depositAddress]]
+                ).read();
+                require(oracleValid, "PCVO: invalid oracle value");
+
+                liquidPcv +=
+                    (oracleValue.asUint256() *
+                        PCVDeposit(depositAddress).balance()) /
+                    1e18;
+            }
+
+            for (uint256 i = 0; i < illiquidVenueLength; i++) {
+                address depositAddress = illiquidVenues.at(i);
+                (Decimal.D256 memory oracleValue, bool oracleValid) = IOracle(
+                    tokenToOracle[depositToToken[depositAddress]]
+                ).read();
+                require(oracleValid, "PCVO: invalid oracle value");
+
+                illiquidPcv +=
+                    (oracleValue.asUint256() *
+                        PCVDeposit(depositAddress).balance()) /
+                    1e18;
+            }
+
+            totalPcv = liquidPcv + illiquidPcv;
+        }
+    }
+
+    /// ------------- PCV Deposit Only API -------------
+
+    /// @notice update the cumulative and last updated times
+    /// only callable by a liquid pcv deposit
+    /// this allows for lazy evaluation of the TWAPCV
+    /// @param pcvDelta the amount of PCV change in the venue
+    function updateLiquidBalance(int256 pcvDelta)
+        external
+        onlyVoltRole(VoltRoles.LIQUID_PCV_DEPOSIT_ROLE)
+    {
+        address token = PCVDeposit(msg.sender).balanceReportedIn();
+        IOracle oracle = IOracle(tokenToOracle[token]);
+        (Decimal.D256 memory oracleValue, bool oracleValid) = oracle.read();
+        require(oracleValid, "PCVO: invalid oracle value");
+        int256 usdPcvDelta = (int256(oracleValue.asUint256()) * pcvDelta) /
+            1e18;
+
+        _updateLiquidBalance(usdPcvDelta);
+        _afterActionHook();
+    }
+
+    /// @notice update the cumulative and last updated times
+    /// only callable by a liquid pcv deposit
+    /// this allows for lazy evaluation of the TWAPCV
+    /// @param pcvDelta the amount of PCV change in the venue
+    function updateIlliquidBalance(int256 pcvDelta)
+        external
+        onlyVoltRole(VoltRoles.ILLIQUID_PCV_DEPOSIT_ROLE)
+    {
+        address token = PCVDeposit(msg.sender).balanceReportedIn();
+        IOracle oracle = IOracle(tokenToOracle[token]);
+        (Decimal.D256 memory oracleValue, bool oracleValid) = oracle.read();
```

Comment:
This setup should allow the protocol to arbitrarily change the value of the LP tokens in this deposit.

Reply:
not if the oracle is manipulation resistant, is that a suggestion or an issue you're trying to point at ?

Reply:
The behavior I'd like to allow is if there is a loss in an illiquid venue, we should be able to mark those assets down within this oracle framework, not by removing the deposit, but by changing the price oracle for that deposit specifically.

Say losses are sustained in Maple, we can't change the USDC oracle to mark USDC down to 50% of its value because there is also USDC in Morpho-Compound, so we need a clean way to handle this type of edgecase.

My solution to this would be map the pcv deposit address to an oracle instead of mapping the underlying token address to the oracle. This way, we can selectively mark down deposits through governance in a much easier manner.